crates/turborepo-lib/src/process/child.rs

```diff
@@ -0,0 +1,473 @@
+                        }
+                    };
+
+                    {
+                        let mut task_state = task_state.write().await;
+                        *task_state = state;
+                    }
+                }
+                status = child.wait() => {
+                    // the child process exited
+                    let exit_code = status.ok().and_then(|s| s.code());
+                        {
+                            let mut task_state = task_state.write().await;
+                            *task_state = ChildState::Finished(exit_code);
+                        }
+                        exit_tx.send(exit_code).ok();
+
+                }
+            }
+
+            debug!("child process exited");
+        });
+
+        Self {
+            pid,
+            gid,
+            state,
+            exit_channel: exit_rx,
+            stdin: Arc::new(Mutex::new(stdin)),
+            stdout: Arc::new(Mutex::new(stdout)),
+            stderr: Arc::new(Mutex::new(stderr)),
+        }
+    }
+
+    /// Wait for the `Child` to exit, returning the exit code.
+    pub async fn wait(&mut self) -> Option<i32> {
+        self.exit_channel.changed().await.ok()?;
+        *self.exit_channel.borrow()
+    }
+
+    /// Perform a graceful shutdown of the `Child` process.
+    pub async fn stop(&mut self) {
+        {
+            let state = self.state.read().await;
+            let child = match Self::child_channel(&state) {
+                Some(child) => child,
+                None => return,
+            };
+
+            // if this fails, it's because the channel is dropped (toctou)
+            // we can just ignore it
+            child.stop().await.ok();
+        }
+
+        self.wait().await;
+    }
+
+    /// Kill the `Child` process immediately.
+    pub async fn kill(&mut self) {
+        {
+            let rw_lock_read_guard = self.state.read().await;
+            let child = match Self::child_channel(&rw_lock_read_guard) {
+                Some(child) => child,
+                None => return,
+            };
+
+            // if this fails, it's because the channel is dropped (toctou)
+            // we can just ignore it
+            child.kill().await.ok();
+        }
+
+        self.wait().await;
+    }
+
+    fn pid(&self) -> Option<u32> {
+        self.pid
+    }
+
+    pub fn stdin(&mut self) -> Option<tokio::process::ChildStdin> {
+        self.stdin.lock().unwrap().take()
+    }
+
+    pub fn stdout(&mut self) -> Option<tokio::process::ChildStdout> {
+        self.stdout.lock().unwrap().take()
+    }
+
+    pub fn stderr(&mut self) -> Option<tokio::process::ChildStderr> {
+        self.stderr.lock().unwrap().take()
+    }
+
+    /// Get a channel for interacting with the child process.
+    fn child_channel(state: &ChildState) -> Option<&ChildCommandChannel> {
+        match state {
+            ChildState::Running(child) => Some(child),
+            _ => None,
```

Comment:
Nit: probably better to enumerate all of the variants so that we get an error if we ever add one.